librad/src/git/storage.rs

```diff
@@ -161,13 +175,40 @@ impl Storage {
                 };
                 self.backend
                     .refname_to_id(&format!("refs/namespaces/{}/refs/{}", namespace, branch))
-                    .and_then(|tip| {
-                        Ok(tip == commit.id()
-                            || self.backend.graph_descendant_of(tip, commit.id())?)
+                    .and_then(|tip| self.has_tag_or_commit(tip, &object))
+                    .or_matches(is_not_found_err, || {
+                        tracing::debug!("ref not found");
+                        Ok(false)
                     })
-                    .or_matches(is_not_found_err, || Ok(false))
             })
-            .or_matches(is_not_found_err, || Ok(false))
+            .or_matches(is_not_found_err, || {
+                tracing::debug!("object not found");
+                Ok(false)
+            })
+    }
+
+    fn has_tag_or_commit(
```

Comment:
@kim I didn't exactly use the `ObjectType` but instead decided to use `as_object`. I believe this covers the requirements you pointed out